llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp

```diff
@@ -0,0 +1,649 @@
+        if (Instr->getOperand(0).getMBB() == Merge)
+          Instr->getOperand(0).setMBB(NewMerge);
+      }
+      MergeToHeader.emplace(NewMerge, MBB);
+    }
+    return modified;
+  }
+
+  // Modifies the CFG to make sure the same block is not both a continue target,
+  // and a merge target.
+  bool splitMergeAndContinueBlocks(MachineFunction &MF) {
+    MachineDominatorTree MDT(MF);
+    std::vector<MachineBasicBlock *> toProcess;
+    visit(MF,
+          [&toProcess](MachineBasicBlock *MBB) { toProcess.push_back(MBB); });
+
+    auto ContinueBlocks = getContinueBlocks(MF);
+    bool modified = false;
+    for (auto *MBB : toProcess) {
+      MachineBasicBlock *Merge = nullptr;
+      MachineBasicBlock *Continue = nullptr;
+      MachineInstr *MI = nullptr;
+      if (!getMergeInstructionTargets(MBB, &MI, &Merge, &Continue))
+        continue;
+
+      if (ContinueBlocks.count(Merge) == 0)
+        continue;
+
+      // This blocks' merge is another block's continue.
+      modified = true;
+      MachineBasicBlock *NewMerge = splitMergeBlock(MDT, MF, *Merge, *MBB);
+      MI->getOperand(0).setMBB(NewMerge);
+    }
+    return modified;
+  }
+
+  // Sorts basic blocks by dominance to respect the SPIR-V spec.
+  bool sortBlocks(MachineFunction &MF) {
+    MachineDominatorTree MDT(MF);
+
+    std::unordered_map<MachineBasicBlock *, size_t> Order;
+    size_t Index = 0;
+    visit(MF,
+          [&Order, &Index](MachineBasicBlock *MBB) { Order[MBB] = Index++; });
+
+    auto Comparator = [&Order](MachineBasicBlock &LHS, MachineBasicBlock &RHS) {
+      return Order[&LHS] < Order[&RHS];
+    };
+
+    MF.sort(Comparator);
+    // FIXME: need to check if the order changed. Maybe if the comparator
+    // returns false once, it did?
+    return true;
+  }
+
+  // In some cases, divergence is allowed without any OpSelectionMerge
+  // instruction because paths only one path doesn't end-up to the parent's
+  // selection/loop construct merge. Example:
+  //          A
+  //         / \
+  //        B<--C
+  //        \    \
+  //         \    D
+  //          \  /
+  //           E
+  // In this case, E is A's merge node.
+  // Previous steps marked C as a selection construct header because it has an
+  // out-degree of 2. But the thread divergence state cannot merge earlier: due
+  // to this triangle configuration, there is no earlier merge node than E. This
+  // means we are still in the same selection construct, hence don't require a
+  // new OpSelectionMerge.
+  bool removeSuperfluousSelectionHeaders(MachineFunction &MF) {
+    MachineDominatorTree MDT(MF);
+
+    bool modified = false;
+    for (MachineBasicBlock &MBB : MF) {
+      MachineInstr *MI = getMergeInstruction(MBB);
+      if (MI == nullptr)
+        continue;
+      // This doesn't apply to block targeted by a back-edge.
+      if (MI->getOpcode() == SPIRV::OpLoopMerge)
+        continue;
+
+      size_t dominated_count = 0;
```

Comment:
Please start [variable names with upper case letter and follow with camel case](https://llvm.org/docs/CodingStandards.html#id44).